src/app/courses-page/courses-page/search/search.component.ts

```diff
@@ -8,9 +8,18 @@ import { Component, Output, EventEmitter } from '@angular/core';
 export class SearchComponent {
   @Output() search = new EventEmitter<string>();
 
-  searchQuery : string = '';
+  public searchQuery : string = '';
+  private timeoutId : any;
 
-  onSearch() {
-    this.search.emit(this.searchQuery);
+  onSearch(event) {
+    const searchQuery = event.target.value;
+    this.searchQuery = searchQuery;
+    clearTimeout(this.timeoutId);
+
+    if (searchQuery.length > 3) {
+      this.timeoutId = setTimeout(() => {
```

Comment:
it is better to separate debounce and real searching, do you remember such thing like ngModelOptions.debounce in angularJS? There is similar in angular, but more advanced with usage of observable https://stackoverflow.com/questions/43998536/debounce-to-get-a-value-input-in-angular2